Development practices/Background and scheduled jobs.md

```diff
@@ -0,0 +1,59 @@
+
+- send emails
+- send push notifications
+- execute transactions
+- generate PDFs
+- interact with 3rd party APIs
+- do other long-running requests, eg. updating database indices (Algolia)
+
+In this case, the web request can take up much more of your resources, which will impact the speed of page response times and potentially result in page timeouts during high traffic.
+
+## Background jobs
+The solution to these problems is to extract the part that does the heavy-lifting and execute it asynchronously.
+
+The extracted part is called a *background job* which can be scheduled or executed at the time of the incoming request in another thread.
+
+To get background jobs to work, we need the following infrastructure:
+
+- database to store jobs - [redis](https://github.com/redis/redis-rb)
+- processor that executes jobs - [sidekiq](https://github.com/mperham/sidekiq)
+
+Configuring Sidekiq to work with Redis can be found [here](https://github.com/mperham/sidekiq/wiki/Using-Redis).
+After you [setup](https://github.com/mperham/sidekiq/wiki/Getting-Started) a worker (background job) you can hook it up wherever is required and choose when it's going to be executed.
+
+### Error handling
+Another great benefit for using background jobs is improving fault tolerance, especially if you have an API that communicates with another.
+
+We can never be 100% sure that the dependent API works OK, which could cause errors in the database or 500 HTTP errors.
+
+Sidekiq background jobs provide [options](https://github.com/mperham/sidekiq/wiki/Advanced-Options#workers) that are defined as class methods which can be used to improve error handling.
+
+### Active Job
+Active Job is a framework for declaring jobs and making them run on a variety of [queuing backends](https://guides.rubyonrails.org/active_job_basics.html#starting-the-backend).
+
+To setup Sidekiq as a queuing library for Active Job read [this Wiki](https://github.com/mperham/sidekiq/wiki/Active-Job).
+
+## Scheduled jobs
+Sometimes a service needs to be run recurrently and not in a HTTP request lifecycle:
+
+- every N (seconds/minutes/hours/days)
+- on a certain day of every month
+- on a certain hour of every day
+- etc...
+
+This is where [sidekiq-scheduler](https://github.com/moove-it/sidekiq-scheduler) comes in handy.
+It also provides the `cron` option for specifying when the service will be called if you prefer or require it.
+
+## Sidekiq Monitoring
+API's usually require a dashboard interface for the admin users. It makes sense to also include a monitoring tool that has a lot of useful info in regards to background jobs.
+
+You can use [built-in methods](https://github.com/mperham/sidekiq/wiki/Monitoring#devise) in your `config/routes.rb` file to enable this feature.
+#### _NOTE:_
+Because of nginx, you'll have to copy the Sidekiq [web assets](https://github.com/mperham/sidekiq/tree/master/web/assets) to the `public/sidekiq` directory.
```

Comment:
This is not because of nginx, but because of our nginx setup.

```
location ~ \.(js|css|png|jpg|jpeg|gif|ico|xml|swf|flv|eot|ttf|woff|pdf|xls|htc|svg)$ {
    add_header Pragma "public";
    add_header Cache-Control "public, must-revalidate, proxy-revalidate";
    expires   30d;
}
location ~* \.(eot|otf|ttf|woff|woff2?)$ {
   add_header Access-Control-Allow-Origin *;
   add_header Pragma "public";
   add_header Cache-Control "public, must-revalidate, proxy-revalidate";
   expires    30d;
}
```

we have this setup, which says all assets should be served as a static files, and are searched inside current/public folder. but since sidekiq assets are inside a gem, nginx wont find them...

We can either do what you are suggesting.
we can ask devops to try to do this: https://github.com/mperham/sidekiq/wiki/FAQ#issue-with-static-asset-serving
or we could create a link to sidekiq assets on deploy https://github.com/mperham/sidekiq/issues/260

Reply:
I like the [asset link solution](https://github.com/mperham/sidekiq/issues/260#issuecomment-10277420). Should I test it out and write a bit more in this chapter? We could use it always like this imho...feels cleaner than copying the asset files from older projects into `public/sidekiq`

Reply:
I changed this section of the chapter with both solutions and added a bit more info regarding nginx